pkg/trace/agent/agent.go

```diff
@@ -205,48 +205,54 @@ func (a *Agent) Process(t *api.Trace) {
 			traceutil.SetMeta(root, tagContainersTags, t.ContainerTags)
 		}
 	}
+	pt := ProcessedTrace{
+		Trace:         t.Spans,
+		WeightedTrace: stats.NewWeightedTrace(t.Spans, root),
+		Root:          root,
+		Env:           a.conf.DefaultEnv,
+		Sublayers:     make(map[*pb.Span][]stats.SublayerValue),
+	}
+
+	sampledSpans, sampled := a.sample(priority, hasPriority, ts, pt)
 
 	// Figure out the top-level spans and sublayers now as it involves modifying the Metrics map
 	// which is not thread-safe while samplers and Concentrator might modify it too.
 	traceutil.ComputeTopLevel(t.Spans)
 
 	subtraces := stats.ExtractSubtraces(t.Spans, root)
-	sublayers := make(map[*pb.Span][]stats.SublayerValue)
 	for _, subtrace := range subtraces {
 		subtraceSublayers := stats.ComputeSublayers(subtrace.Trace)
-		sublayers[subtrace.Root] = subtraceSublayers
-		stats.SetSublayersOnSpan(subtrace.Root, subtraceSublayers)
+		pt.Sublayers[subtrace.Root] = subtraceSublayers
+		if sampled {
+			stats.SetSublayersOnSpan(subtrace.Root, subtraceSublayers)
+		}
 	}
 
-	pt := ProcessedTrace{
-		Trace:         t.Spans,
-		WeightedTrace: stats.NewWeightedTrace(t.Spans, root),
-		Root:          root,
-		Env:           a.conf.DefaultEnv,
-		Sublayers:     sublayers,
-	}
 	if tenv := traceutil.GetEnv(t.Spans); tenv != "" {
 		// this trace has a user defined env.
 		pt.Env = tenv
 	}
 
-	if priority >= 0 {
-		a.sample(ts, pt)
-	}
-
 	a.Concentrator.In <- &stats.Input{
 		Trace:     pt.WeightedTrace,
 		Sublayers: pt.Sublayers,
 		Env:       pt.Env,
 	}
+
+	if sampled {
+		a.Out <- sampledSpans
+	}
 }
 
 // sample decides whether the trace will be kept and extracts any APM events
 // from it.
-func (a *Agent) sample(ts *info.TagStats, pt ProcessedTrace) {
-	var ss writer.SampledSpans
+func (a *Agent) sample(priority sampler.SamplingPriority, hasPriority bool, ts *info.TagStats, pt ProcessedTrace) (*writer.SampledSpans, bool) {
```

Comment:
This became quite nasty. Would it make sense to absorb this information into `ProcessedTrace`?